tests/components/ecobee/test_climate.py

```diff
@@ -335,3 +389,18 @@ async def test_set_fan_mode_auto(thermostat, data):
     data.ecobee.set_fan_mode.assert_has_calls(
         [mock.call(1, "auto", "nextTransition", holdHours=None)]
     )
+
+
+async def test_turn_aux_heat_on(thermostat, data):
```

Comment:
We should not use the `thermostat` fixture in new code as it is returning the actual Entity which is interacting with any integration details and not allowed:

https://developers.home-assistant.io/docs/development_testing?_highlight=testing#writing-tests-for-integrations

Reply:
Call services via the core service registry [hass.services](https://github.com/home-assistant/core/blob/4cce724473233d4fb32c08bd251940b1ce2ba570/homeassistant/core.py#L1133). instead

Reply:
This integration is pretty old so the existing test code has some examples of what not to do and needs some refactoring.

Reply:
That's kind of what I figured, and why I changed the other added test, but I wasn't sure how to convert this turn_aux_heat_off and on to using (hass) instead. 

Any tips? Create a Thermostat() object and run the code there? or do I need to do it within accessing hass states/services?

Reply:
Oops - I commented without seeing your other few responses.

Reply:
okay @bdraco 
I assume I do something like this:
```
    await setup_platform(hass, CLIMATE_DOMAIN)
    await hass.services.async_call(
        CLIMATE_DOMAIN,
        "turn_aux_heat_on()",
        {ATTR_ENTITY_ID: ENTITY_ID},
        blocking=True,
    )
    await hass.async_block_till_done()
```
But what do I put for the service? And just for clarity, is it always needed to call services, opposed to creating the object and mocking/running the functions directly? If so, I may need to change some test on my own integration pr.

Reply:
Here are a few examples

https://github.com/home-assistant/core/blob/dev/tests/components/demo/test_climate.py#L297

Reply:
Okay - so using those examples, and registering a service in ecobee/climate.py, I then have the following test:
This test is not functional, as I am not sure the correct way to do the next part. turn_aux_heat_on calls set_hvac_mode in the api, which then sends a http request to an api and sets the hvac mode. If I mocked/set the value that the api call does, I'm not sure if that is a valid test anymore as it would almost certainly pass. I would appreciate your insight.

```
async def test_turn_aux_heat_on(hass):
    """Test when aux heat is set on.  This must change the HVAC mode."""
    mock_get_thermostat = mock.Mock()
    mock_get_thermostat.return_value = GENERIC_THERMOSTAT_INFO_WITH_HEATPUMP
    with mock.patch("pyecobee.Ecobee.get_thermostat", mock_get_thermostat):
        await setup_platform(hass, CLIMATE_DOMAIN)
    mock_response = mock.Mock()
    with mock.patch("pyecobee.Ecobee._request_with_refresh", mock_response):
        await hass.services.async_call(
            CLIMATE_DOMAIN,
            ecobee.SERVICE_TURN_AUX_HEAT_ON,
            {ATTR_ENTITY_ID: ENTITY_ID},
            blocking=True,
        )
    assert mock_get_thermostat.return_value.settings == ecobee.ECOBEE_AUX_HEAT_ONLY
```

Reply:
It's generally easier for me to give you feedback if you push the code and than I can make comments inline about how to improve/fix as it can get confusing quickly when we have a long comment thread

Reply:
okay I pushed the current state of my code

Reply:
https://github.com/bdraco/home-assistant/commit/e5768322266195aa1eb6a1a2715426b4609e733b

Hopefully this will get you started.  I did not build out a fully mocked `Ecobee` object which will be needed to get the integration to setup

Reply:
You should be able to copy some of the tests from the recent PR to this integration to use as a template.  I know there is a lot of legacy code in the tests. If you are still stuck I'll try to find some time to dig in some more.

Reply:
Sounds good. I have some time today that I set aside to try to figure this out. Hopefully all goes well.

Reply:
@bdraco Think I got it! Let me know what you think

Reply:
No rush, but did you have a chance to look @bdraco? I should have more time to work on it tonight if anything is wrong, so I figured I would check